[WIP] upgrade net::ssh version

kbrock opened the pull request:
Upgrade net ssh version to a version that properly handles remote trust

Currently, we require a user to manually add it before running commands like `fix_auth --fetch-key`:

```sh
eval `ssh-agent -s`
ssh-add
```

This is a common scenario and not a hack. But it does require a certain level of knowledge of linux and ssh that may be beyond most users.

With the upgrade of the net::ssh gem, the admin no longer needs to manage the ssh trust chain.

@bdunne or @Fryguy : I am not able to judge the danger of this change. And requiring the user to properly setup their trust chain may be valid. The only breaking changes I saw around this is requiring a different gem dependency.

I did not test this. I could not get the appliance console cli to pickup this change.
Does look like we have a number of gems that depend upon 4.2, not sure if this is pegged because of an incompatibility or if it has bigger reasons.

addresses https://github.com/ManageIQ/manageiq/issues/20091
fixed https://bugzilla.redhat.com/show_bug.cgi?id=1824306

djberg96 commented:
There is a `remote_trust` option you can pass, but it's false by default. Does that not work?

I wouldn't think upgrading would be an issue, but I would do some general tinkering just to make sure.

kbrock commented:
Thanks so much @djberg96 

I did not play with that option. I'm guessing version 6 has that value set to true by default. just passing this option seems less intrusive

djberg96 commented:
Oops, sorry, there is no such option. I think I meant `remember_host`.

kbrock commented:
@djberg96 thank you for the backup

I tried `:remote_trust` and `:forward_agent` (I know it was a stretch) but they did not help me.

There is some documentation for `start` but did not see `:remote_trust`. [docs](http://net-ssh.github.io/net-ssh/Net/SSH.html#method-c-start)

thinking about deferring this update and asking users to follow standard linux protocols

Fryguy commented:
Do we have any tests in other repos that exercise this code?  I'm fine with merging, but would feel more comfortable if one of the existing usage paths are verified.

kbrock commented:
Ok, so I think when generating keys the system is generating an elliptical cipher. This is understandable by ssh-agent and net-ssh 6.1.

We can:

- ask user to use an rsa key (they probably thought this is what they were generating - it used to be the default) - trying to debug this
- ask user to use `ssh-agent`
- upgrade net-ssh from 4.2 to 6.1
  - upgrade net-ssh
    - manageiq-gems-pending
    - linux_admin
    - manageiq-ssh-util
  - adding native gem `ed25519`

demo: (`Gemfile` and `hello_host.rb`) - for some reason a single file did not work for me.

```ruby
source "https://rubygems.org"

if ENV["SSH"] != "old"
  puts "netssh 6.1"

  # requires:
  # dnf install redhat-rpm-config
  gem "net-ssh", "~> 6.1"
  gem "ed25519", "1.2.4"
else
  puts "netssh 4.2"

  # already installed on machine:
  # dnf install libsodium
  # do not need:
  # gem "rbnacl-libsodium"

  #gem "manageiq-ssh-util", "~>0.1.1"
  gem "net-ssh", "~> 4.2.0"
  gem "rbnacl", "4.0.2"
  #gem "rbnacl-libsodium"
end
gem "bcrypt_pbkdf"
```

```ruby
require 'net/ssh'

hostarg = ARGV[0] || 'root@10.0.2.15'
unless (hostname = hostarg.split("@").last) and (user = hostarg.split("@").first)
  puts "usage: program user@host"
  exit
end

puts "connecting to #{hostname} as #{user}"
Net::SSH.start(hostname, user) do |ssh|
  puts ssh.exec!("hostname")
end
```

```
dnf install redhat-rpm-config # probably not needed if building at rpm build time
SSH=new BUNDLE_GEMFILE=./Gemfile bundle
SSH=new BUNDLE_GEMFILE=./Gemfile bundle exec ruby ./hello_host.rb
# works
SSH=old BUNDLE_GEMFILE=./Gemfile bundle
SSH=old BUNDLE_GEMFILE=./Gemfile bundle exec ruby ./hello_host.rb
# key was 385 bytes (Expected 32) (RbNaCl::LengthError)
```
